File: 0095-unique-binary-search-trees-ii/95.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def build_all_trees(self, start: int, end: int) -> List[Optional[TreeNode]]:
        if start > end:
            return [None]
        
        all_trees = []
        for root_val in range(start, end + 1):
            left_trees = self.build_all_trees(start, root_val - 1)
            right_trees = self.build_all_trees(root_val + 1, end)

            for l in left_trees:
                for r in right_trees:
                    root = TreeNode(root_val)
                    root.left = l
                    root.right = r
                    all_trees.append(root)

        return all_trees

    def generateTrees(self, n: int) -> List[Optional[TreeNode]]:
        """
        Approach:
            1. [Recursion] -- T.C: O(n * C(n)), S.C: O(n * C(n)); C(n): nth Catalan Number
                - Each number in [1..n] is choosen as root, and all combinations of left/right subtrees are built recursively.
        """

        return self.build_all_trees(1, n)
        

File: 0095-unique-binary-search-trees-ii/95.c
/** Definition for a binary tree node.
 * struct TreeNode {
 *   int              val;
 *   struct TreeNode *left;
 *   struct TreeNode *right;
 * };
 */

// Dynamic array to store trees
typedef struct {
  struct TreeNode **data;
  int               size;
  int               capacity;
} TreeList;

void initTreeList(TreeList *list) {
  list->capacity = 8;
  list->size     = 0;
  list->data     = (struct TreeNode **)malloc(sizeof(struct TreeNode *) * list->capacity);
}

void addTree(TreeList *list, struct TreeNode *tree) {
  if (list->size == list->capacity) {
    list->capacity *= 2;
    list->data = realloc(list->data, sizeof(struct TreeNode *) * list->capacity);
  }
  list->data[list->size++] = tree;
}

struct TreeNode *newNode(int val) {
  struct TreeNode *node = malloc(sizeof(struct TreeNode));
  node->val             = val;
  node->left            = NULL;
  node->right           = NULL;
  return node;
}

/**
 * Note: The returned array must be malloced, assume caller calls free().
 */
TreeList buildAllTrees(int start, int end) {
  TreeList list;
  initTreeList(&list);

  // Base Case: no nodes in this range
  if (start > end) {
    addTree(&list, NULL);
    return list;
  }

  for (int i = start; i <= end; i++) {
    TreeList leftTrees  = buildAllTrees(start, i - 1);
    TreeList rightTrees = buildAllTrees(i + 1, end);

    for (int l = 0; l < leftTrees.size; l++) {
      for (int r = 0; r < rightTrees.size; r++) {
        struct TreeNode *root = newNode(i);
        root->left            = leftTrees.data[l];
        root->right           = rightTrees.data[r];
        addTree(&list, root);
      }
    }

    free(leftTrees.data);
    free(rightTrees.data);
  }

  return list;
}

struct TreeNode **generateTrees(int n, int *returnSize) {
  /* Approach:
     1. [Recursion] -- T.C: O(n * C(n)), S.C: O(n * C(n)); C(n): nth Catalan Number
  */
  TreeList result = buildAllTrees(1, n);
  *returnSize     = result.size;
  return result.data;
}
